### DISCUSSION TITLE
$extends behaving like $use

### DISCUSSION THREAD

**Discussion by liam-taly** [prisma/prisma#23908] [Q&A] [answered]
### Question

I need to capture the timings (metrics) for the prisma queries and I can easily do it with $use. How can I modify it to use $extends?

BTW, the metrics returned by $on is not what I am looking for. I need the metrics per prisma query and not per SQL query.

```js
  prismaTemp.$use(async (params, next) => {
    const startTime = Date.now();
    const result = await next(params);
    const endTime = Date.now();
    const duration = endTime - startTime;

    const args = JSON.stringify(convertBigIntToString(params.args));

    console.log({
      MetricName: 'QueryDuration',
      Dimensions: [
        {
          Name: 'model',
          Value: params.model ?? 'Unknown'
        },
        {
          Name: 'action',
          Value: params.action
        },
        {
          Name: 'args',
          Value: args
        },
        {
          Name: 'runInTransaction',
          Value: params.runInTransaction
        },
      ],
      Unit: 'Milliseconds',
      Value: duration
    })

    return result;
  })
```
Any help appreciated

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @liam-taly 👋 

Thank you for raising this question.

You can use the [`$extends`](https://www.prisma.io/docs/orm/prisma-client/client-extensions) method to create an extended Prisma Client and add custom methods to it.

An example on how to use an extended Prisma client for query logging is demonstrated in this [Prisma examples repository](https://github.com/prisma/prisma-client-extensions/blob/main/query-logging/README.md) and you can see the actual extensions implemented [here](https://github.com/prisma/prisma-client-extensions/blob/main/query-logging/script.ts). 

For capturing the timings for each Prisma query, you might want to look into the [OpenTelemetry tracing](https://www.prisma.io/docs/concepts/components/prisma-client/opentelemetry-tracing) and [Metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) features, which provide detailed insights into performance and how Prisma interacts with the database. Please note that these features are still in preview and might not be stable for production use.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by liam-taly** [reply]
Than you ludralph!

**Reply by ludralph** [reply]
You're welcome! I'm glad I could assist you. If you're interested in exploring more about Prisma's offerings, I'd recommend checking out [Accelerate](https://www.prisma.io/docs/accelerate/what-is-accelerate) and [Pulse](https://www.prisma.io/docs/pulse/what-is-pulse).

Accelerate is a global database cache with scalable connection pooling, and Pulse is a managed Change Data Capture (CDC) service that enables real-time database subscriptions. Both of these can be integrated into your application via the [Platform Console](https://console.prisma.io/).
